rustc_codegen_llvm: add support for writing summary bitcode

durin42 opened the pull request:
Typical uses of ThinLTO don't have any use for this as a standalone file, but distributed ThinLTO uses this to make the linker phase more efficient. With clang you'd do something like `clang -flto=thin -fthin-link-bitcode=foo.indexing.o -c foo.c` and then get both foo.o (full of bitcode) and foo.indexing.o (just the summary or index part of the bitcode). That's then usable by a two-stage linking process that's more friendly to distributed build systems like bazel, which is why I'm working on this area.

I talked some to @teresajohnson about naming in this area, as things seem to be a little confused between various blog posts and build systems. "bitcode index" and "bitcode summary" tend to be a little too ambiguous, and she tends to use "thin link bitcode" and "minimized bitcode" (which matches the descriptions in LLVM). Since the clang option is thin-link-bitcode, I went with that to try and not add a new spelling in the world.

Per @dtolnay, you can work around the lack of this by using `lld --thinlto-index-only` to do the indexing on regular .o files of bitcode, but that is a bit wasteful on actions when we already have all the information in rustc and could just write out the matching minimized bitcode. I didn't test that at all in our infrastructure, because by the time I learned that I already had this patch largely written.

durin42 committed:
rustc_codegen_llvm: add support for writing summary bitcode

Typical uses of ThinLTO don't have any use for this as a standalone
file, but distributed ThinLTO uses this to make the linker phase more
efficient. With clang you'd do something like `clang -flto=thin
-fthin-link-bitcode=foo.indexing.o -c foo.c` and then get both foo.o
(full of bitcode) and foo.indexing.o (just the summary or index part of
the bitcode). That's then usable by a two-stage linking process that's
more friendly to distributed build systems like bazel, which is why I'm
working on this area.

I talked some to @teresajohnson about naming in this area, as things
seem to be a little confused between various blog posts and build
systems. "bitcode index" and "bitcode summary" tend to be a little too
ambiguous, and she tends to use "thin link bitcode" and "minimized
bitcode" (which matches the descriptions in LLVM). Since the clang
option is thin-link-bitcode, I went with that to try and not add a new
spelling in the world.

Per @dtolnay, you can work around the lack of this by using `lld
--thinlto-index-only` to do the indexing on regular .o files of
bitcode, but that is a bit wasteful on actions when we already have all
the information in rustc and could just write out the matching minimized
bitcode. I didn't test that at all in our infrastructure, because by the
time I learned that I already had this patch largely written.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9164417833/job/25195682360) [(plain)](https://github.com/rust-lang/rust/commit/74ad3bdf3625ee6ad6242bbd4447c7bd5d01a368/checks/25195682360/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.6s done
#16 DONE 32.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
    | ^^^^^^^^^^^^^^^^^^^
467 |     Bitcode,
468 |     ThinLinkBitcode,
    |     --------------- not covered
    = note: the matched value is of type `OutputType`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
223 ~             OutputType::Mir | OutputType::Metadata | OutputType::Exe | OutputType::DepInfo => {},
224 +             OutputType::ThinLinkBitcode => todo!()

For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_codegen_cranelift` (lib) due to 1 previous error
Build completed unsuccessfully in 0:03:16
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9165403273/job/25198774540) [(plain)](https://github.com/rust-lang/rust/commit/c34660e4851a837c46eab53dcc3c5ed3dff44cb2/checks/25198774540/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 28.7s done
#17 DONE 33.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling tempfile v3.7.1
error[E0046]: not all trait items implemented, missing: `thin_link_data`
   --> src/lib.rs:334:1
    |
334 | impl ThinBufferMethods for ThinBuffer {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `thin_link_data` in implementation
    |
    = help: implement the missing item: `fn thin_link_data(&self) -> &[u8] { todo!() }`
For more information about this error, try `rustc --explain E0046`.
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
Build completed unsuccessfully in 0:03:23
  local time: Mon May 20 21:52:29 UTC 2024
```

</details>

Nadrieril commented:
r? compiler

lcnr commented:
r? compiler

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9176421923/job/25231658165) [(plain)](https://github.com/rust-lang/rust/commit/01c1ffd340cbd500eb49e2277bffdc2d16968eec/checks/25231658165/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.9s done
#16 DONE 40.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
diff of stderr:

5 .intel_syntax noprefix
6 ^
7 error: unknown directive
+ .intel_syntax noprefix
+ error: unknown directive
8    |
9 note: instantiated into assembly here
9 note: instantiated into assembly here
10   --> <inline asm>:1:1

The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.arm/inline-syntax.arm.stderr
To only update this specific test, also pass `--test-args asm/inline-syntax.rs`


error in revision `arm`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/asm/inline-syntax.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "arm" "--check-cfg" "cfg(FALSE,x86_64,arm,arm_llvm_18)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.arm" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.arm/auxiliary" "--target" "armv7-unknown-linux-gnueabihf"
--- stderr -------------------------------
error: unknown directive
.intel_syntax noprefix
^
---
.intel_syntax noprefix
^
error: unknown directive
   |
note: instantiated into assembly here
  --> <inline asm>:1:1
   |
LL | .intel_syntax noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:35:15
   |
   |
LL |         asm!(".intel_syntax noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
LL |     .intel_syntax noprefix
   |     ^

error: unknown directive
error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:39:15
   |
LL |         asm!(".intel_syntax aaa noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .intel_syntax aaa noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:43:15
   |
   |
LL |         asm!(".att_syntax noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .att_syntax noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:47:15
   |
   |
LL |         asm!(".att_syntax bbb noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .att_syntax bbb noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:51:15
   |
   |
LL |         asm!(".intel_syntax noprefix; nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .intel_syntax noprefix; nop

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:58:13
   |
   |
LL |             .intel_syntax noprefix
   |             ^
   |
note: instantiated into assembly here
  --> <inline asm>:2:13
LL |             .intel_syntax noprefix
   |             ^

error: aborting due to 7 previous errors
```

</details>

durin42 commented:
Latest CI run failed with `      [ui] tests/ui/asm/inline-syntax.rs#arm` - which I shouldn't have been able to influence at all with this change.

nnethercote commented:
This is enabled with `--emit=thin-link-bitcode`, correct?

The code in the PR looks ok to me, though LTO experts might have opinions -- @bjorn3? @mw?

But I think that adding a new output type is a significant enough change that it should be done via the [Major Change Proposal](https://forge.rust-lang.org/compiler/mcp.html) mechanism.

bjorn3 commented:
> But I think that adding a new output type is a significant enough change that it should be done via the [Major Change Proposal](https://forge.rust-lang.org/compiler/mcp.html) mechanism.

https://github.com/rust-lang/compiler-team/issues/735

nnethercote commented:
Ok, thanks for the extra info. @bjorn3 do you want to give the final approval here? You know what's going on here much better than I do.

bjorn3 commented on compiler/rustc_codegen_llvm/src/back/write.rs:
```diff
@@ -725,12 +728,31 @@ pub(crate) unsafe fn codegen(
-                    .generic_activity_with_arg("LLVM_module_codegen_emit_bitcode", &*module.name);
-                if let Err(err) = fs::write(&bc_out, data) {
-                    dcx.emit_err(WriteBytecode { path: &bc_out, err });
+                {
```
Why this extra block? It spans the entirety of the enclosing if statement.

durin42 committed:
cleanup: remove leftover extra block

This was needed in an older version of this patch, but never got edited
out when it became obsolete.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9195201026/job/25290444406) [(plain)](https://github.com/rust-lang/rust/commit/ad0ef0c1bee62b77b448d46ef135ad1f3a60a1ad/checks/25290444406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 30.5s done
#16 DONE 33.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
diff of stderr:

5 .intel_syntax noprefix
6 ^
7 error: unknown directive
+ .intel_syntax noprefix
+ error: unknown directive
8    |
9 note: instantiated into assembly here
9 note: instantiated into assembly here
10   --> <inline asm>:1:1

The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.arm/inline-syntax.arm.stderr
To only update this specific test, also pass `--test-args asm/inline-syntax.rs`


error in revision `arm`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/asm/inline-syntax.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "arm" "--check-cfg" "cfg(FALSE,x86_64,arm,arm_llvm_18)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.arm" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.arm/auxiliary" "--target" "armv7-unknown-linux-gnueabihf"
--- stderr -------------------------------
error: unknown directive
.intel_syntax noprefix
^
---
.intel_syntax noprefix
^
error: unknown directive
   |
note: instantiated into assembly here
  --> <inline asm>:1:1
   |
LL | .intel_syntax noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:35:15
   |
   |
LL |         asm!(".intel_syntax noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
LL |     .intel_syntax noprefix
   |     ^

error: unknown directive
error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:39:15
   |
LL |         asm!(".intel_syntax aaa noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .intel_syntax aaa noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:43:15
   |
   |
LL |         asm!(".att_syntax noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .att_syntax noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:47:15
   |
   |
LL |         asm!(".att_syntax bbb noprefix", "nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .att_syntax bbb noprefix

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:51:15
   |
   |
LL |         asm!(".intel_syntax noprefix; nop");
   |
note: instantiated into assembly here
note: instantiated into assembly here
  --> <inline asm>:1:2
   |
LL |     .intel_syntax noprefix; nop

error: unknown directive
##[error]  --> /checkout/tests/ui/asm/inline-syntax.rs:58:13
   |
   |
LL |             .intel_syntax noprefix
   |             ^
   |
note: instantiated into assembly here
  --> <inline asm>:2:13
LL |             .intel_syntax noprefix
   |             ^

error: aborting due to 7 previous errors
```

</details>

durin42 commented:
I have confirmed that, somehow, the llvm-17 regression here is real and seemingly caused by my change. I'll investigate soon.

durin42 committed:
thinlto: only build summary file if needed

If we don't do this, some versions of LLVM (at least 17, experimentally)
will double-emit some error messages, which is how I noticed this. Given
that it seems to be costing some extra work, let's only request the
summary bitcode production if we'll actually bother writing it down,
otherwise skip it.

durin42 committed:
cleanup: standardize on summary over index in names

I did this in the user-facing logic, but I noticed while fixing a minor
defect that I had missed it in a few places in the internal details.

durin42 commented:
Fixed the llvm 17 failure. Sorry for the force-push, that ended up being hard to avoid once I got into the weeds debugging.

I also cleaned up a couple of variable names so things are generally consistent in using `summary` instead of `index`.

durin42 committed:
cleanup: run rustfmt

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9213548486/job/25347717359) [(plain)](https://github.com/rust-lang/rust/commit/a0581b5b7fe6b9add5c1f0fe90a9f2cc8249b01e/checks/25347717359/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 31.9s done
#16 DONE 37.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling object v0.30.4
   Compiling gccjit_sys v0.1.0
   Compiling gccjit v2.0.0
   Compiling tempfile v3.7.1
error[E0050]: method `prepare_thin` has 1 parameter but the declaration in trait `prepare_thin` has 2
    |
    |
421 |     fn prepare_thin(_module: ModuleCodegen<Self::Module>) -> (String, Self::ThinBuffer) {
    |
    |
    = note: `prepare_thin` from trait: `fn(ModuleCodegen<<Self as WriteBackendMethods>::Module>, bool) -> (std::string::String, <Self as WriteBackendMethods>::ThinBuffer)`
For more information about this error, try `rustc --explain E0050`.
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
Build completed unsuccessfully in 0:03:25
  local time: Thu May 23 19:17:01 UTC 2024
```

</details>

durin42 committed:
rustc_codegen_gcc: fix changed method signature

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit cfe3f77f9dd83bceb35e9685c4690b7b1a5be1b0 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 cfe3f77f9dd83bceb35e9685c4690b7b1a5be1b0 -->
<!-- homu: {"type":"Approved","sha":"cfe3f77f9dd83bceb35e9685c4690b7b1a5be1b0","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
